src/treelearner/serial_tree_learner.cpp

```diff
@@ -147,10 +148,11 @@ void SerialTreeLearner::ResetConfig(const Config* config) {
   constraints_.reset(LeafConstraintsBase::Create(config_, config_->num_leaves));
 }
 
-Tree* SerialTreeLearner::Train(const score_t* gradients, const score_t *hessians) {
+Tree* SerialTreeLearner::Train(const score_t* gradients, const score_t *hessians, bool is_constant_hessian, const Json& forced_split_json) {
```

Comment:
I am quite confused about the changes of `forced_split_json` and `is_constant_hessian`. Why these changes are needed?

Reply:
Many of these new parameters are needed for the CUDA subclass of TreeLearner.

Reply:
I think I remove the 'is_constant_hessian_' in PR https://github.com/microsoft/LightGBM/pull/2827 , and put it to the `share_state_`.  Is there any particular reason to add it back?

Reply:
and `forced_split_json` is refactoring in PR https://github.com/microsoft/LightGBM/pull/2867

Reply:
I am working on fixing the "forced_split_json" and "is_constant_hessian" parameter issues, and plan to drop that fixed code soon (assuming it passes my regression tests).